#**Inicio**

In [359]:
import numpy as np
import pandas as pd

In [360]:
df=pd.read_csv('/content/drive/MyDrive/arquivos_aula/heart_tratadoa.Csv', sep=';', encoding='utf-8')

In [361]:
df.head(10)

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA      140.0        289.0          0     Normal    172   
1   49   F           NAP      160.0        180.0          0     Normal    156   
2   37   M           ATA      130.0        283.0          0         ST     98   
3   48   F           ASY      138.0        214.0          0     Normal    108   
4   54   M           NAP      150.0        195.0          0     Normal    122   
5   39   M           NAP      120.0        339.0          0     Normal    170   
6   45   F           ATA      130.0        237.0          0     Normal    170   
7   54   M           ATA      110.0        208.0          0     Normal    142   
8   37   M           ASY      140.0        207.0          0     Normal    130   
9   48   F           ATA      120.0        284.0          0     Normal    120   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  
5              N      0.0       Up             0  
6              N      0.0       Up             0  
7              N      0.0       Up             0  
8              Y      1.5     Flat             1  
9              N      0.0       Up             0

In [362]:
df.shape

(918, 12)

#**Transformaçâo as variáveis Categorias Nominais em Variáveis categorias Ordinais**

In [372]:
df2=pd.DataFrame.copy(df)

In [373]:
df2.Sex.replace({'M':0,'F':1},inplace=True)
df2.ChestPainType.replace({'TA':0,'ATA':1,'NAP':2,'ASY':3},inplace=True)
df2.RestingECG.replace({'Normal':0,'ST':1,'LVH':3},inplace=True)
df2.ExerciseAngina.replace({'N':0,'Y':1},inplace=True)
df2.ST_Slope.replace({'Up':0,'Flat':1,'Down':2},inplace=True)

In [374]:
df2.dtypes

Age                 int64
Sex                 int64
ChestPainType       int64
RestingBP         float64
Cholesterol       float64
FastingBS           int64
RestingECG          int64
MaxHR               int64
ExerciseAngina      int64
Oldpeak           float64
ST_Slope            int64
HeartDisease        int64
dtype: object

In [375]:
df2.head(20)

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0    40    0              1      140.0        289.0          0           0   
1    49    1              2      160.0        180.0          0           0   
2    37    0              1      130.0        283.0          0           1   
3    48    1              3      138.0        214.0          0           0   
4    54    0              2      150.0        195.0          0           0   
5    39    0              2      120.0        339.0          0           0   
6    45    1              1      130.0        237.0          0           0   
7    54    0              1      110.0        208.0          0           0   
8    37    0              3      140.0        207.0          0           0   
9    48    1              1      120.0        284.0          0           0   
10   37    1              2      130.0        211.0          0           0   
11   58    0              1      136.0        164.0          0           1   
12   39    0              1      120.0        204.0          0           0   
13   49    0              3      140.0        234.0          0           0   
14   42    1              2      115.0        211.0          0           1   
15   54    1              1      120.0        273.0          0           0   
16   38    0              3      110.0        196.0          0           0   
17   43    1              1      120.0        201.0          0           0   
18   60    0              3      100.0        248.0          0           0   
19   36    0              1      120.0        267.0          0           0   

    MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0     172               0      0.0         0             0  
1     156               0      1.0         1             1  
2      98               0      0.0         0             0  
3     108               1      1.5         1             1  
4     122               0      0.0         0             0  
5     170               0      0.0         0             0  
6     170               0      0.0         0             0  
7     142               0      0.0         0             0  
8     130               1      1.5         1             1  
9     120               0      0.0         0             0  
10    142               0      0.0         0             0  
11     99               1      2.0         1             1  
12    145               0      0.0         0             0  
13    140               1      1.0         1             1  
14    137               0      0.0         0             0  
15    150               0      1.5         1             0  
16    166               0      0.0         1             1  
17    165               0      0.0         0             0  
18    125               0      1.0         1             1  
19    160               0      3.0         1             1

Age = idade (anos)

Sex = sexo (0=M; 1=F)

Chest Pain Type = tipo de dor no peito (0=TA:angina típica; 1=ATA:angina atípica; 2=NAP:dor não anginosa; 3=ASY:assintomático)

Resting BP = pressão sanguínea em repouso (mmHg)

Cholesterol = colesterol sérico (mg/dl)

Fasting BS = açúcar no sangue em jejum (mg/dl)
          0: Fasting BS < 120 mg/dl (não diabético)
          1: Fasting BS >= 120 mg/dl, (diabético)

Resting ECG = eletrocardiograma em repouso (0=Normal; 1=ST:Anormalidade da onda ST-T; 2=LVH: Hipertrofia ventricular esquerda)

Max HR = frequência cardíaca máxima

Exercise Angina = Angina induzida por exercício (0=Não; 1=Sim)

Old Peak = Depressão de ST induzida por exercício em relação ao repouso

ST_Slope = Inclinação do segmento ST (0=UP; 1=Flat; 2=Down)

Heart Disease = Doença cardíaca (0=Não possui doença cardíaca; 1=Possui doença cardíaca)

**negrito**

In [376]:
previsores=df2.iloc[:,0:11].values

In [377]:
print(previsores)

[[40.   0.   1.  ...  0.   0.   0. ]
 [49.   1.   2.  ...  0.   1.   1. ]
 [37.   0.   1.  ...  0.   0.   0. ]
 ...
 [57.   0.   3.  ...  1.   1.2  1. ]
 [57.   1.   1.  ...  0.   0.   1. ]
 [38.   0.   2.  ...  0.   0.   0. ]]


In [378]:
previsores.shape

(918, 11)

In [379]:
alvo=df2.iloc[:,11].values

In [380]:
print(alvo)

[0 1 0 1 0 0 0 0 1 0 0 1 0 1 0 0 1 0 1 1 0 0 0 1 0 0 0 0 0 0 1 0 1 1 0 0 1
 0 0 0 0 1 0 0 1 0 0 0 0 1 1 1 0 0 0 0 1 1 0 1 0 0 0 1 0 0 0 0 1 0 1 0 1 0
 1 0 1 0 0 1 0 0 1 0 1 1 1 0 1 0 0 0 0 1 0 1 0 0 0 0 1 0 1 1 1 0 0 0 0 0 0
 1 0 0 0 1 1 1 0 1 1 0 0 1 0 0 0 0 0 0 0 1 1 1 0 1 0 0 1 1 1 1 1 0 1 0 0 0
 0 1 0 0 0 0 0 1 1 0 1 0 1 1 0 0 0 1 1 0 0 0 0 0 0 0 1 1 1 0 0 0 1 0 1 0 0
 1 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 1 1 1 0 0 1 0 1 0 0 0 1 1
 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 1 1 1 0 1 1 0 1 0 1 1 1 1 1 1 0 0 1 0 0 0 0
 0 0 0 1 1 1 0 1 0 1 0 0 0 1 0 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 1
 1 1 1 1 1 0 1 1 1 1 1 0 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0
 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 0 1 1 0 0 1 0 1 1 0 1 1 1 1 0 1 1 0 0 1 1 1 0 1 1 1 1
 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 0 1 1 1 0 1 0 1 0 1 0 1 1 1 1 0 1 0 1 1 1 1
 1 1 1 1 1 0 1 0 1 1 1 1 

In [381]:
alvo.shape

(918,)

In [382]:
df2.describe()

Age         Sex  ChestPainType   RestingBP  Cholesterol  \
count  918.000000  918.000000     918.000000  918.000000   918.000000   
mean    53.510893    0.210240       2.251634  132.540894   244.635389   
std      9.432617    0.407701       0.931031   17.989932    53.318029   
min     28.000000    0.000000       0.000000   80.000000    85.000000   
25%     47.000000    0.000000       2.000000  120.000000   214.000000   
50%     54.000000    0.000000       3.000000  130.000000   244.635389   
75%     60.000000    0.000000       3.000000  140.000000   267.000000   
max     77.000000    1.000000       3.000000  200.000000   603.000000   

        FastingBS  RestingECG       MaxHR  ExerciseAngina     Oldpeak  \
count  918.000000  918.000000  918.000000      918.000000  918.000000   
mean     0.233115    0.808279  136.809368        0.404139    0.887364   
std      0.423046    1.176958   25.460334        0.490992    1.066570   
min      0.000000    0.000000   60.000000        0.000000   -2.600000   
25%      0.000000    0.000000  120.000000        0.000000    0.000000   
50%      0.000000    0.000000  138.000000        0.000000    0.600000   
75%      0.000000    1.000000  156.000000        1.000000    1.500000   
max      1.000000    3.000000  202.000000        1.000000    6.200000   

         ST_Slope  HeartDisease  
count  918.000000    918.000000  
mean     0.638344      0.553377  
std      0.607056      0.497414  
min      0.000000      0.000000  
25%      0.000000      0.000000  
50%      1.000000      1.000000  
75%      1.000000      1.000000  
max      2.000000      1.000000


**Padronização (utiliza a média e o desvio padrão como referência).**

**Normalização (utiliza os valores máximo e mínimo como referência).**

In [383]:
from sklearn.preprocessing import StandardScaler

In [384]:
previsores_esc=StandardScaler().fit_transform(previsores)

In [385]:
previsores_esc

array([[-1.4331398 , -0.51595242, -1.34508565, ..., -0.8235563 ,
        -0.83243239, -1.05211381],
       [-0.47848359,  1.93816322, -0.27042192, ..., -0.8235563 ,
         0.10566353,  0.59607813],
       [-1.75135854, -0.51595242, -1.34508565, ..., -0.8235563 ,
        -0.83243239, -1.05211381],
       ...,
       [ 0.37009972, -0.51595242,  0.80424181, ...,  1.21424608,
         0.29328271,  0.59607813],
       [ 0.37009972,  1.93816322, -1.34508565, ..., -0.8235563 ,
        -0.83243239,  0.59607813],
       [-1.64528563, -0.51595242, -0.27042192, ..., -0.8235563 ,
        -0.83243239, -1.05211381]])

In [386]:
previsores_df=pd.DataFrame(previsores_esc)

In [387]:
previsores_df.describe()

0             1             2             3             4   \
count  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02   
mean  -1.083616e-16  7.740117e-18  1.935029e-17 -7.914270e-16 -1.876978e-16   
std    1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00   
min   -2.706015e+00 -5.159524e-01 -2.419749e+00 -2.922164e+00 -2.995655e+00   
25%   -6.906294e-01 -5.159524e-01 -2.704219e-01 -6.974863e-01 -5.748916e-01   
50%    5.188098e-02 -5.159524e-01  8.042418e-01 -1.413168e-01  0.000000e+00   
75%    6.883185e-01 -5.159524e-01  8.042418e-01  4.148527e-01  4.196855e-01   
max    2.491558e+00  1.938163e+00  8.042418e-01  3.751870e+00  6.724929e+00   

                 5             6             7             8             9   \
count  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02   
mean  -3.096047e-17  3.096047e-17  4.953675e-16 -3.870058e-18  1.238419e-16   
std    1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00   
min   -5.513413e-01 -6.871266e-01 -3.018469e+00 -8.235563e-01 -3.271482e+00   
25%   -5.513413e-01 -6.871266e-01 -6.605778e-01 -8.235563e-01 -8.324324e-01   
50%   -5.513413e-01 -6.871266e-01  4.678968e-02 -8.235563e-01 -2.695748e-01   
75%   -5.513413e-01  1.629842e-01  7.541571e-01  1.214246e+00  5.747115e-01   
max    1.813758e+00  1.863206e+00  2.561874e+00  1.214246e+00  4.983762e+00   

                 10  
count  9.180000e+02  
mean  -7.740117e-17  
std    1.000545e+00  
min   -1.052114e+00  
25%   -1.052114e+00  
50%    5.960781e-01  
75%    5.960781e-01  
max    2.244270e+00

## **Codificação de variáveis categóricas**
### **LabelEncoder: transformação de variáveis categóricas em Numéricas**

In [388]:
from sklearn.preprocessing import LabelEncoder

In [389]:
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA      140.0        289.0          0     Normal    172   
1   49   F           NAP      160.0        180.0          0     Normal    156   
2   37   M           ATA      130.0        283.0          0         ST     98   
3   48   F           ASY      138.0        214.0          0     Normal    108   
4   54   M           NAP      150.0        195.0          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [390]:
previsores_2=df.iloc[:,0:11].values
previsores_2

array([[40, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       [49, 'F', 'NAP', ..., 'N', 1.0, 'Flat'],
       [37, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       ...,
       [57, 'M', 'ASY', ..., 'Y', 1.2, 'Flat'],
       [57, 'F', 'ATA', ..., 'N', 0.0, 'Flat'],
       [38, 'M', 'NAP', ..., 'N', 0.0, 'Up']], dtype=object)

In [391]:
previsores_2[:,1]=LabelEncoder().fit_transform(previsores_2[:,1])

In [392]:
previsores_2

array([[40, 1, 'ATA', ..., 'N', 0.0, 'Up'],
       [49, 0, 'NAP', ..., 'N', 1.0, 'Flat'],
       [37, 1, 'ATA', ..., 'N', 0.0, 'Up'],
       ...,
       [57, 1, 'ASY', ..., 'Y', 1.2, 'Flat'],
       [57, 0, 'ATA', ..., 'N', 0.0, 'Flat'],
       [38, 1, 'NAP', ..., 'N', 0.0, 'Up']], dtype=object)

In [393]:
previsores_2[:,2]=LabelEncoder().fit_transform(previsores_2[:,2])
previsores_2[:,6]=LabelEncoder().fit_transform(previsores_2[:,6])
previsores_2[:,8]=LabelEncoder().fit_transform(previsores_2[:,8])
previsores_2[:,10]=LabelEncoder().fit_transform(previsores_2[:,10])

In [394]:
previsores_2

array([[40, 1, 1, ..., 0, 0.0, 2],
       [49, 0, 2, ..., 0, 1.0, 1],
       [37, 1, 1, ..., 0, 0.0, 2],
       ...,
       [57, 1, 0, ..., 1, 1.2, 1],
       [57, 0, 1, ..., 0, 0.0, 1],
       [38, 1, 2, ..., 0, 0.0, 2]], dtype=object)

In [395]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

In [396]:
ColumnTransformer?

Parâmetros ColumnTransformer
- name: nome dado a transformação.
- transformer: tipo de estimador (OneHotEncoder).
- columns: colunas que serão transformadas.
- remainder: o que acontecerá com o restante das colunas não relacionadas: 
1) drop = exclui as outras colunas.
2) passthrough = mantém as outras colunas. drop é default.  
- sparse_threshold: parâmetro de classificação de matrizes esparsas. default é 0.3
- n_jobs: número de trabalhos a serem executados em paralelo. default é nenhum
- transformer_weights: definição de pesos aos transformadores.
-verbose: default é False. se for True a execução é apresentada na tela.

In [397]:
previsores_3=ColumnTransformer(transformers=[('Onehot',OneHotEncoder(),[1,2,6,8,10])],remainder='passthrough').fit_transform(previsores_2)

In [398]:
previsores_3

array([[0.0, 1.0, 0.0, ..., 0, 172, 0.0],
       [1.0, 0.0, 0.0, ..., 0, 156, 1.0],
       [0.0, 1.0, 0.0, ..., 0, 98, 0.0],
       ...,
       [0.0, 1.0, 1.0, ..., 0, 115, 1.2],
       [1.0, 0.0, 0.0, ..., 0, 174, 0.0],
       [0.0, 1.0, 0.0, ..., 0, 173, 0.0]], dtype=object)

In [399]:
previsores_3.shape

(918, 20)

In [400]:
previsores_3df=pd.DataFrame(previsores_3)
previsores_3df.head()

0    1    2    3    4    5    6    7    8    9    10   11   12   13  14  \
0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  40   
1  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0  49   
2  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  1.0  37   
3  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  48   
4  0.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  54   

      15     16 17   18   19  
0  140.0  289.0  0  172  0.0  
1  160.0  180.0  0  156  1.0  
2  130.0  283.0  0   98  0.0  
3  138.0  214.0  0  108  1.5  
4  150.0  195.0  0  122  0.0

## **Escalonamento**

In [401]:
from sklearn.preprocessing import StandardScaler

In [402]:
previsores_3esc=StandardScaler().fit_transform(previsores_3)

In [403]:
previsores_3esc

array([[-0.51595242,  0.51595242, -1.08413811, ..., -0.55134134,
         1.38292822, -0.83243239],
       [ 1.93816322, -1.93816322, -1.08413811, ..., -0.55134134,
         0.75415714,  0.10566353],
       [-0.51595242,  0.51595242, -1.08413811, ..., -0.55134134,
        -1.52513802, -0.83243239],
       ...,
       [-0.51595242,  0.51595242,  0.9223917 , ..., -0.55134134,
        -0.85706875,  0.29328271],
       [ 1.93816322, -1.93816322, -1.08413811, ..., -0.55134134,
         1.4615246 , -0.83243239],
       [-0.51595242,  0.51595242, -1.08413811, ..., -0.55134134,
         1.42222641, -0.83243239]])

In [404]:
previsores_3df=pd.DataFrame(previsores_3esc)
previsores_3df

0         1         2         3         4         5         6   \
0   -0.515952  0.515952 -1.084138  2.075177 -0.532838 -0.229679 -0.507478   
1    1.938163 -1.938163 -1.084138 -0.481887  1.876744 -0.229679 -0.507478   
2   -0.515952  0.515952 -1.084138  2.075177 -0.532838 -0.229679 -0.507478   
3    1.938163 -1.938163  0.922392 -0.481887 -0.532838 -0.229679 -0.507478   
4   -0.515952  0.515952 -1.084138 -0.481887  1.876744 -0.229679 -0.507478   
..        ...       ...       ...       ...       ...       ...       ...   
913 -0.515952  0.515952 -1.084138 -0.481887 -0.532838  4.353909 -0.507478   
914 -0.515952  0.515952  0.922392 -0.481887 -0.532838 -0.229679 -0.507478   
915 -0.515952  0.515952  0.922392 -0.481887 -0.532838 -0.229679 -0.507478   
916  1.938163 -1.938163 -1.084138  2.075177 -0.532838 -0.229679  1.970528   
917 -0.515952  0.515952 -1.084138 -0.481887  1.876744 -0.229679 -0.507478   

           7         8         9         10        11        12        13  \
0    0.814275 -0.490449  0.823556 -0.823556 -0.271448 -1.002181  1.150674   
1    0.814275 -0.490449  0.823556 -0.823556 -0.271448  0.997824 -0.869056   
2   -1.228087  2.038947  0.823556 -0.823556 -0.271448 -1.002181  1.150674   
3    0.814275 -0.490449 -1.214246  1.214246 -0.271448  0.997824 -0.869056   
4    0.814275 -0.490449  0.823556 -0.823556 -0.271448 -1.002181  1.150674   
..        ...       ...       ...       ...       ...       ...       ...   
913  0.814275 -0.490449  0.823556 -0.823556 -0.271448  0.997824 -0.869056   
914  0.814275 -0.490449  0.823556 -0.823556 -0.271448  0.997824 -0.869056   
915  0.814275 -0.490449 -1.214246  1.214246 -0.271448  0.997824 -0.869056   
916 -1.228087 -0.490449  0.823556 -0.823556 -0.271448  0.997824 -0.869056   
917  0.814275 -0.490449  0.823556 -0.823556 -0.271448 -1.002181  1.150674   

           14        15        16        17        18        19  
0   -1.433140  0.414853  0.832529 -0.551341  1.382928 -0.832432  
1   -0.478484  1.527192 -1.212922 -0.551341  0.754157  0.105664  
2   -1.751359 -0.141317  0.719935 -0.551341 -1.525138 -0.832432  
3   -0.584556  0.303619 -0.574892 -0.551341 -1.132156  0.574711  
4    0.051881  0.971022 -0.931438 -0.551341 -0.581981 -0.832432  
..        ...       ...       ...       ...       ...       ...  
913 -0.902775 -1.253656  0.363389 -0.551341 -0.188999  0.293283  
914  1.536902  0.637320 -0.968969  1.813758  0.164684  2.357094  
915  0.370100 -0.141317 -2.132437 -0.551341 -0.857069  0.293283  
916  0.370100 -0.141317 -0.162048 -0.551341  1.461525 -0.832432  
917 -1.645286  0.303619 -1.306750 -0.551341  1.422226 -0.832432  

[918 rows x 20 columns]

In [405]:
previsores_3df.describe()

0             1             2             3             4   \
count  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02   
mean  -7.740117e-18  7.740117e-18 -1.548023e-16 -9.288140e-17  6.192094e-17   
std    1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00   
min   -5.159524e-01 -1.938163e+00 -1.084138e+00 -4.818867e-01 -5.328378e-01   
25%   -5.159524e-01  5.159524e-01 -1.084138e+00 -4.818867e-01 -5.328378e-01   
50%   -5.159524e-01  5.159524e-01  9.223917e-01 -4.818867e-01 -5.328378e-01   
75%   -5.159524e-01  5.159524e-01  9.223917e-01 -4.818867e-01 -5.328378e-01   
max    1.938163e+00  5.159524e-01  9.223917e-01  2.075177e+00  1.876744e+00   

                 5             6             7           8             9   \
count  9.180000e+02  9.180000e+02  9.180000e+02  918.000000  9.180000e+02   
mean   7.740117e-18  1.238419e-16  9.288140e-17    0.000000  7.740117e-17   
std    1.000545e+00  1.000545e+00  1.000545e+00    1.000545  1.000545e+00   
min   -2.296787e-01 -5.074783e-01 -1.228087e+00   -0.490449 -1.214246e+00   
25%   -2.296787e-01 -5.074783e-01 -1.228087e+00   -0.490449 -1.214246e+00   
50%   -2.296787e-01 -5.074783e-01  8.142748e-01   -0.490449  8.235563e-01   
75%   -2.296787e-01 -5.074783e-01  8.142748e-01   -0.490449  8.235563e-01   
max    4.353909e+00  1.970528e+00  8.142748e-01    2.038947  8.235563e-01   

                 10            11            12            13            14  \
count  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02   
mean  -7.740117e-18 -7.740117e-18  4.644070e-17 -6.192094e-17 -1.238419e-16   
std    1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00   
min   -8.235563e-01 -2.714484e-01 -1.002181e+00 -8.690559e-01 -2.706015e+00   
25%   -8.235563e-01 -2.714484e-01 -1.002181e+00 -8.690559e-01 -6.906294e-01   
50%   -8.235563e-01 -2.714484e-01  9.978237e-01 -8.690559e-01  5.188098e-02   
75%    1.214246e+00 -2.714484e-01  9.978237e-01  1.150674e+00  6.883185e-01   
max    1.214246e+00  3.683942e+00  9.978237e-01  1.150674e+00  2.491558e+00   

                 15            16            17            18            19  
count  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02  9.180000e+02  
mean  -7.856219e-16  2.481675e-15  5.418082e-17  5.108477e-16 -1.548023e-16  
std    1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00  1.000545e+00  
min   -2.922164e+00 -2.995655e+00 -5.513413e-01 -3.018469e+00 -3.271482e+00  
25%   -6.974863e-01 -5.748916e-01 -5.513413e-01 -6.605778e-01 -8.324324e-01  
50%   -1.413168e-01  2.666753e-15 -5.513413e-01  4.678968e-02 -2.695748e-01  
75%    4.148527e-01  4.196855e-01 -5.513413e-01  7.541571e-01  5.747115e-01  
max    3.751870e+00  6.724929e+00  1.813758e+00  2.561874e+00  4.983762e+00

## **RESUMO PRÉ-PROCESSAMENTO**
alvo = variável que se pretende atingir (tem ou não doença cardíaca).

previsores = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas manualmente, sem escalonar.

previsores_esc = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas, escalonada.

previsores2 = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas pelo labelencoder.

previsores3 = conjunto de variáveis previsoras transformadas pelo labelencoder e onehotencoder, sem escalonar.

previsores3_esc = conjunto de variáveis previsoras transformadas pelo labelencoder e onehotencoder escalonada.

## **Redução de dimensionalidade**

Objetivo é selecionar os melhores componentes (atributos) para treinamento do algoritmo, através da análise das correlações entre as variáveis.

### Análise dos Componentes Principais (PCA)

**Seleção de características:** seleciona os melhores atributos e utiliza sem transformações.

**Extração de Características:** Encontra os relacionamentos dos melhores atributos e cria novos atributos.

É um algoritmo de aprendizagem não supervisionada.

Aplica-se em dados linearmente separáveis.

In [406]:
from sklearn.decomposition import PCA

In [407]:
from re import S
#pca=PCA(n_components=8,svd_solver='arpack')
pca=PCA(n_components=4)

In [408]:
previsores_pca=pca.fit_transform(previsores_2)

In [409]:
previsores_pca.shape

(918, 4)

In [410]:
previsores_pca

array([[  44.01234136,  -36.13313394,   10.63630221,   -9.48103316],
       [ -63.99165876,  -13.92452332,   31.68548932,   -5.30865857],
       [  38.53873225,   33.9411749 ,  -12.48715778,  -21.53786541],
       ...,
       [-113.3464858 ,   23.50553174,   -2.51001516,    1.14802677],
       [  -9.11397906,  -35.88381206,    4.80834072,    9.11443821],
       [ -70.0132701 ,  -35.66913385,   12.09887226,  -10.52641614]])

In [411]:
pca.explained_variance_ratio_

array([0.72841104, 0.1719045 , 0.08121586, 0.01767508])

In [412]:
pca.explained_variance_ratio_.sum()

0.999206476034731

### Kernel PCA
É um algoritmo de aprendizagem não supervisionada.

Aplica-se também em dados linearmente não separáveis.
from sklearn.decomposition import KernelPCA
kpca = KernelPCA(n_components=4, kernel='rbf')
previsores_kernel = kpca.fit_transform(previsores2)
previsores_kernel.shape
previsores_kernel

### **Análise do Discriminante Linear (LDA: Linear Discriminant Analysis)**

Algoritmo de aprendizagem supervisionada, pois utiliza a classe como referência para seleção.

Aplicado em situações com muitos atributos previsores e também com o atributo alvo com muitas classes.



In [413]:
from sklearn.decomposition import KernelPCA

In [414]:
kpca=KernelPCA(n_components=4,kernel='rbf')

In [415]:
previsores_Kernel=kpca.fit_transform(previsores_2)

### **Análise do Discriminante Linear (LDA: Linear Discriminant Analysis)**
Algoritmo de aprendizagem supervisionada, pois utiliza a classe como referência para seleção.

Aplicado em situações com muitos atributos previsores e também com o atributo alvo com muitas classes.
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
lda = LinearDiscriminantAnalysis(n_components = 1)
previsores_lda = lda.fit_transform(previsores2, alvo)
previsores_lda
lda.explained_variance_ratio_

In [416]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis 
lda = LinearDiscriminantAnalysis(n_components=1)

In [417]:
previsores_lda=lda.fit_transform(previsores_2,alvo)

In [418]:
previsores_lda

array([[-1.47042594e+00],
       [-1.40849772e+00],
       [-9.68596591e-01],
       [ 5.24976483e-01],
       [-1.45538724e+00],
       [-1.77988847e+00],
       [-2.29164347e+00],
       [-1.26719848e+00],
       [ 1.02744265e+00],
       [-1.80852666e+00],
       [-2.56377252e+00],
       [ 1.08983974e+00],
       [-1.42998021e+00],
       [ 1.02019103e+00],
       [-2.52001678e+00],
       [-7.51690488e-01],
       [-2.71578418e-01],
       [-2.34186361e+00],
       [ 5.09957591e-01],
       [ 4.16533521e-02],
       [-2.89437427e+00],
       [-2.28024293e-01],
       [-2.27124413e+00],
       [ 9.75041059e-01],
       [-1.70761821e+00],
       [-2.06839620e+00],
       [ 1.62868203e+00],
       [-9.64764497e-01],
       [-1.52740493e+00],
       [-1.33371276e+00],
       [-2.88018838e-02],
       [-1.44270127e+00],
       [ 6.44000384e-01],
       [ 3.52720094e-01],
       [-2.24389918e+00],
       [-1.48930961e+00],
       [ 2.46829040e+00],
       [-2.14031598e+00],
       [-1.5

In [419]:
lda.explained_variance_ratio_

array([1.])

## **Salvando as variáveis (atributos)**

In [420]:
import pickle

In [421]:
# Criando arquivo
arq1 = open('heart1_alvo.pkl', 'wb')

In [422]:
# Salvando a variável no arquivo
pickle.dump(alvo, arq1)

In [423]:
# Fechando arquivo
arq1.close()

In [424]:
# Lendo arquivo
arq1 = open('heart1_alvo.pkl', 'rb')

In [425]:
# Chamando a variável
alvo = pickle.load(arq1)

In [426]:
alvo

array([0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,

In [427]:
# Criando arquivos e Salvando as variáveis
arq2=open('heart2_previsores.pkl','wb')
pickle.dump(previsores,arq2)
arq3=open('heart3_previsores_esc.pkl','wb')
pickle.dump(previsores_esc,arq3)
arq4=open('heart4_previsores_2.pkl','wb')
pickle.dump(previsores_2,arq4)
arq5=open('heart5_previsores_3.pkl','wb')
pickle.dump(previsores_3,arq5)
arq6=open('heart6_previsores_3esc.pkl','wb')
pickle.dump(previsores_3esc,arq6)

In [428]:
arq2=open('heart2_previsores.pkl','rb')

In [429]:
previsores=pickle.load(arq2)

In [430]:
previsores

array([[40. ,  0. ,  1. , ...,  0. ,  0. ,  0. ],
       [49. ,  1. ,  2. , ...,  0. ,  1. ,  1. ],
       [37. ,  0. ,  1. , ...,  0. ,  0. ,  0. ],
       ...,
       [57. ,  0. ,  3. , ...,  1. ,  1.2,  1. ],
       [57. ,  1. ,  1. , ...,  0. ,  0. ,  1. ],
       [38. ,  0. ,  2. , ...,  0. ,  0. ,  0. ]])

## **BASE DE TREINO E TESTE**

In [431]:
from sklearn.model_selection import train_test_split

Parâmetros train_test_split:   
- arrays: nomes dos atributos previsores e alvo.   
- test_size: tamanho em porcentagem dos dados de teste. default é none.   
- train_size: tamanho em porcentagem dos dados de treinamento.default é none.  
- random_state: nomeação de um estado aleatório.   
- shuffle: embaralhamento dos dados aleatórios. Associado com o random_state ocorre o mesmo embaralhamento sempre. Default é True.  
- stratify: Possibilidade de dividir os dados de forma estratificada. Default é None (nesse caso é mantido a proporção, isto é, se tem 30% de zeros e 70% de 1 no dataframe, na separação em treinamento e teste se manterá essa proporção).

In [500]:
x_treino, x_teste, y_treino, y_teste = train_test_split(previsores_3esc, alvo, test_size = 0.3, random_state = 6)

In [501]:
x_treino.shape

(642, 20)

In [502]:
x_teste.shape

(276, 20)

In [503]:
y_treino.shape

(642,)

In [504]:
y_teste.shape

(276,)

# **NAIVE BAYES**


Treinamento do algoritmo

In [505]:
from sklearn.naive_bayes import GaussianNB

In [506]:
naive=GaussianNB()
naive.fit(x_treino,y_treino)

GaussianNB()

Avaliação do algoritmo

In [ ]:
previsoes_naive=naive.predict(x_teste)
previsoes_naive

In [ ]:
y_teste

In [509]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [510]:
accuracy_score(y_teste,previsoes_naive )

0.855072463768116

In [511]:
print("Acurácia:%.2f%%"%(accuracy_score(y_teste,previsoes_naive)*100))

Acurácia:85.51%


In [512]:
confusion_matrix(y_teste,previsoes_naive)

array([[ 97,  21],
       [ 19, 139]])

In [ ]:
print(classification_report(y_teste,previsoes_naive))

**Análise dados de treino**

In [515]:
previsoes_treino=naive.predict(x_treino)

In [ ]:
previsoes_treino

In [ ]:
accuracy_score(y_treino,previsoes_treino)

In [524]:
confusion_matrix(y_treino,previsoes_treino)

array([[247,  45],
       [ 51, 299]])

### **Validação Cruzada**

In [519]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [520]:
from pandas.core.frame import np_percentile_argname
# separando os dados em folds
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

In [ ]:
# Criando o modelo
modelo = GaussianNB()
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)
resultado

In [522]:
#Usando media de desfio padrão
print("Acurácia Média: %.2f%%" % (resultado.mean()*100))

Acurácia Média: 84.39%
